src/ModelContextProtocol/Server/McpServer.cs

```diff
@@ -401,7 +414,19 @@ private void SetToolsHandler(McpServerOptions options)
                     do
                     {
                         ListToolsResult extraResults = await originalListToolsHandler(request, cancellationToken).ConfigureAwait(false);
-                        result.Tools.AddRange(extraResults.Tools);
+                        foreach(var tool in extraResults.Tools)
+                        {
+                            resultTools.Add(tool);
+                            if (progressToken is not null)
+                            {
+                                await this.NotifyProgressAsync(new()
+                                {
+                                    ProgressToken = progressToken,
+                                    Progress = resultTools.Count,
+                                    Message = $"Listing extra tool '{tool.Name}'...",
+                                }, cancellationToken).ConfigureAwait(false);
+                            }
+                        }
```

Comment:
Instead of one notification per tool, how about one per batch? The distance between tools in the batch is zero, so getting a distinct progress update for each seems unnecessary.